storage/innobase/fsp/fsp0fsp.cc

```diff
@@ -3577,6 +3579,238 @@ dberr_t fsp_sys_tablespace_validate()
 }
 #endif /* UNIV_DEBUG */
 
+/** Store the inode information which basically stores
+the page and offset */
+struct inodeInfo
+{
+private:
+  std::map<uint32_t, std::set<uint16_t>> m_used_inodes;
+public:
+  void insert(uint32_t page, uint16_t offset)
+  {
+    m_used_inodes[page].insert(offset);
+  }
+
+  void insert_seg(byte *inode)
+  {
+    m_used_inodes[mach_read_from_4(inode + 4)].insert(
+      mach_read_from_2(inode + 8));
+  }
```

Comment:
Is there a reason why we use `insert` instead of `emplace`? Would `unordered_map` work instead of `map`?

In the common formatting style, lines should never be broken right after an opening parenthesis `(` or `[`. You can break the line after `.` for example.

Reply:
Shouldn’t we report an error if something had already been registered in `m_used_inodes`? That is, if multiple entries point to the same file segment, we should treat the system tablespace as corrupted and refuse to free any file segments or to shrink the system tablespace.